core/ecschema-metadata/src/test/Metadata/Class.test.ts

```diff
@@ -31,6 +31,51 @@ describe("ECClass", () => {
       schema = new Schema(new SchemaContext(), "TestSchema", "ts", 1, 0, 0);
     });
 
+    it("checks if properties are overridden correctly", async () => {
+      const schemaJson = {
+        $schema: ECSchemaNamespaceUris.SCHEMAURL3_2_JSON,
+        name: "TestSchema",
+        version: "1.2.3",
+        alias: "ts",
+        items: {
+          TestBase: {
+            schemaItemType: "EntityClass",
+            properties: [
+              {
+                type: "PrimitiveProperty",
+                typeName: "string",
+                name: "PrimProp",
+                label: "BaseProp",
+              },
+            ],
+          },
+          TestClass: {
+            schemaItemType: "EntityClass",
+            baseClass: "TestSchema.TestBase",
+            properties: [
+              {
+                type: "PrimitiveProperty",
+                typeName: "string",
+                name: "PrimProp",
+                label: "DerivedProp",
+              },
+            ],
+          },
+        },
+      };
+      schema = await Schema.fromJson(schemaJson, new SchemaContext());
+      const testClass = await schema.getItem("TestClass", EntityClass);
+      const testBase = await schema.getItem("TestBase", EntityClass);
+      assert.isDefined(testClass);
+      assert.isDefined(testBase);
+      const testClassPrimProp = await testClass!.getProperty("PrimProp", true);
+      const testBasePrimProp = await testBase!.getProperty("PrimProp", true);
```

Comment:
I don't know if that test setup is any helpful for us:
Since both classes have a property with that name, it would successfully look it up from the class __property_ field and the cache wouldn't even be build. Also if we exclude the inherited properties (2nd argument), the property cache wouldn't even be built either as it would return undefined if the property can't be looked up in the first check.

I guess what we have to do is to add another class which doesn't have own properties, this derives from the _TestClass_, so if we lookup for our _PrimProp_ (including inherited properties), it's forced to build a property cache and we can see whether all works as designed.

Reply:
Since we have _sinon_ available in this package, I'd also propose to add a spy on the _buildPropertyCache_ method, to ensure it has been called.